mlc_llm/transform/smoothquant.py

```diff
@@ -0,0 +1,245 @@
+
+
+@tvm.transform.module_pass(opt_level=0, name="SmoothQuantOpConverter")
+class SmoothQuantOpConverter:
+    def __init__(self, op_name: str) -> None:
+        self.op_name = op_name
+
+    def transform_module(self, mod: tvm.IRModule, ctx: tvm.transform.PassContext) -> tvm.IRModule:
+        attrs = {"mode": "identity"}
+        data = wildcard()
+        scale = wildcard()
+        pattern = is_op("relax.annotate.smooth")(data, scale).has_attr(attrs)
+
+        def rewriter(_, matchings):
+            kind = matchings[pattern].attrs.kind
+            return R.smooth(matchings[data], matchings[scale], kind=kind, mode=self.op_name)
+
+        new_mod = tvm.IRModule()
+        for gv, func in mod.functions.items():
+            if isinstance(func, relax.Function):
+                new_mod[gv] = rewrite_call(pattern, rewriter, func)
+            else:
+                new_mod[gv] = func
+        return new_mod
+
+
+@tvm.transform.module_pass(opt_level=0, name="SmoothQuantLegalizer")
+class SmoothQuantLegalizer:
+    """
+    Pass that converts matmul(fp16, fp16) -> quantize + matmul(int8, int8) + dequantize.
+    """
+    def __init__(self, adtype="int8", wdtype="int8"):
+        self.dtype_act = adtype
+        self.dtype_weight = wdtype
+
+    def transform_module(self, mod: tvm.IRModule, ctx: tvm.transform.PassContext) -> tvm.IRModule:
+        attrs = {"mode": "quantize"}
+        act_scale = wildcard()
+        w_scale = wildcard()
+        lhs_sm = is_op("relax.annotate.smooth")(wildcard(), act_scale).has_attr(attrs)
+        rhs_sm = is_op("relax.annotate.smooth")(wildcard(), w_scale).has_attr(attrs)
+        permute = is_op("relax.permute_dims")(rhs_sm)
+        pattern = is_op("relax.matmul")(lhs_sm, permute)
+
+        def rewriter(_, matchings):
+            def _make_quantize(call: tvm.relax.Call, out_dtype: str):
+                min_value = tvm.tir.min_value(out_dtype)
+                max_value = tvm.tir.max_value(out_dtype)
+                data = R.round(R.divide(call.args[0], call.args[1]))
+                return R.astype(R.clip(data, min_value, max_value), dtype=out_dtype)
+            
+            def _make_dequantize(
+                call: tvm.relax.Call,
+                scale1: tvm.relax.Constant,
+                scale2: tvm.relax.Constant,
+                out_dtype: str,
+            ):
+                if out_dtype == "float32":
+                    return R.multiply(R.astype(call, dtype=out_dtype), R.multiply(scale1, scale2))
+                else:
+                    assert out_dtype == "float16"
+                    min_value = tvm.tir.min_value(out_dtype)
+                    max_value = tvm.tir.max_value(out_dtype)
+                    dq_scale = R.multiply(R.astype(scale1, "float32"), R.astype(scale2, "float32"))
+                    out = R.multiply(R.astype(call, dtype="float32"), dq_scale)
+                    return R.astype(R.clip(out, min_value, max_value), dtype=out_dtype)
```

Comment:
This manual quantize and dequantize should be replaced by Relax "QNN dialect" in the near future.

Reply:
Strongly agree! In my opinion QNN dialect is very desirable in Relax.